core/src/memory/binary_page_memory_resource.cpp

```diff
@@ -0,0 +1,348 @@
+         * allocator, and then look for that new page.
+         */
+        if (cand == nullptr) {
+            allocate_upstream(goal);
+
+            cand = find_free_page(goal);
+        }
+
+        /*
+         * If there is still no candidate, something has gone wrong and we
+         * cannot recover.
+         */
+        if (cand == nullptr) {
+            throw std::bad_alloc();
+        }
+
+        /*
+         * If the page is split (but its children are all free), we will first
+         * need to unsplit it.
+         */
+        if (cand->state == page_state::SPLIT) {
+            cand->unsplit();
+        }
+
+        /*
+         * Keep splitting the page until we have reached our target size.
+         */
+        while (cand->size > goal) {
+            cand->split();
+            cand = cand->left.get();
+        }
+
+        /*
+         * Mark the page as occupied, then return the address.
+         */
+        cand->state = page_state::OCCUPIED;
+
+        return cand->addr;
+    }
+
+    void binary_page_memory_resource::do_deallocate(
+        void * p,
+        std::size_t,
+        std::size_t
+    ) {
+        page * cand = nullptr;
+
+        /*
+         * We will use this stack to perform a depth-first search of the pages
+         * in our binary trees.
+         *
+         * TODO: This would be much more efficient with a hashmap.
+         */
+        std::stack<page *> rem;
+
+        for (std::unique_ptr<page> & p : m_pages) {
+            rem.push(p.get());
+        }
+
+        /*
+         * Iterate over the pages in our tree.
+         */
+        while (!rem.empty()) {
+            page * c = rem.top();
+            rem.pop();
+
+            if (c->addr == p) {
+                /*
+                 * If we have found the target node, we're done.
+                 */
+                cand = c;
+                break;
+            } else if (c->state == page_state::SPLIT) {
+                /*
+                 * If the page is split, we need to add its children to the
+                 * stack.
+                 */
+                rem.push(c->left.get());
+                rem.push(c->right.get());
+            }
+        }
+
+        /*
+         * If we have found the target, just mark it as vacant. There is no need
+         * to issue a deallocation upstream.
+         */
+        if (cand != nullptr) {
+            cand->free();
+        }
+    }
+
+    bool binary_page_memory_resource::do_is_equal(
+        const memory_resource & other
+    ) const noexcept {
+        /*
+         * All malloc resources are equal to each other, because they have no
+         * internal state. Of course they have a shared underlying state in the
+         * form of the underlying C library memory manager, but that is not
+         * relevant for us.
+         */
+        const binary_page_memory_resource * c;
+        c = dynamic_cast<const binary_page_memory_resource *>(&other);
+
+        return this == c;
```

Comment:
:thinking: What's the purpose of this check in the end? I was thinking of this as a test of memory type, and not so much as a test of the exact memory resource class. But I might be thinking about it incorrectly...

What I'm trying to get at is that I thought that in the "high level" memory resources we would compare the "low level" memory resources that they are using. Like:

```c++
return m_upstream->is_equal( *( c->m_upstream ) );
```

But I'm honestly not sure what the correct behaviour should be here... :confused:

Reply:
The C++ standard requires `do_is_equal` to return true if and only if the one memory resource can free memory allocated with the other and vice versa. It's possible to have two of these memory resources with completely different internal states, so I figured this would be a good-enough solution to check that equality constraint.

Reply:
:thinking: Okay, I'm fine with that.

But in that case I think we could spare ourselves the dynamic cast all together, and just do a pointer value comparison. Wouldn't

```c++
return ( this == &other );
```

work without anything else?